File: chkit_python/src/chkit/core/validate.py
"""Validation rules for schema definitions."""

from __future__ import annotations

import re
from collections.abc import Iterable
from typing import Final

from chkit.core.canonical import definition_key
from chkit.core.codec import canonicalize_codec, is_general_codec, is_raw_codec
from chkit.core.key_clause import is_plain_column_reference, normalize_key_columns
from chkit.core.model import (
    ChxValidationError,
    ColumnDefinition,
    DictionaryDefinition,
    MaterializedViewDefinition,
    SchemaDefinition,
    TableDefinition,
    ValidationIssue,
    ValidationIssueCode,
)
from chkit.core.projection import is_index_projection, normalize_projection_index


def _push(
    issues: list[ValidationIssue],
    definition: SchemaDefinition,
    code: ValidationIssueCode,
    message: str,
) -> None:
    issues.append(
        ValidationIssue(
            code=code,
            kind=definition.kind,
            database=definition.database,
            name=definition.name,
            message=message,
        )
    )


def _validate_column_codec(
    definition: TableDefinition, column: ColumnDefinition, issues: list[ValidationIssue]
) -> None:
    if column.codec is None:
        return
    steps = canonicalize_codec(column.codec)
    if len(steps) == 0:
        _push(
            issues,
            definition,
            "codec_chain_empty",
            f'Table {definition.database}.{definition.name} column "{column.name}" '
            f"codec chain is empty; provide at least one codec or omit the field",
        )
        return

    general_count = 0
    general_index = -1
    for i, step in enumerate(steps):
        if is_raw_codec(step):
            continue
        if is_general_codec(step):
            general_count += 1
            general_index = i

    if general_count > 1:
        _push(
            issues,
            definition,
            "codec_chain_multiple_general",
            f'Table {definition.database}.{definition.name} column "{column.name}" '
            f"codec chain has more than one general codec; only one general codec is "
            f"allowed at the end of a chain",
        )
        return

    if len(steps) > 1 and general_count == 1 and general_index != len(steps) - 1:
        _push(
            issues,
            definition,
            "codec_chain_must_end_with_general",
            f'Table {definition.database}.{definition.name} column "{column.name}" '
            f"codec chain must end with a general codec "
            f"(NONE, LZ4, LZ4HC, ZSTD, T64, GCD, ALP)",
        )


def _validate_indexes(definition: TableDefinition, issues: list[ValidationIssue]) -> None:
    index_seen: set[str] = set()
    for index in definition.indexes or []:
        if index.name in index_seen:
            _push(
                issues,
                definition,
                "duplicate_index_name",
                f'Table {definition.database}.{definition.name} '
                f'has duplicate index name "{index.name}"',
            )
            continue
        index_seen.add(index.name)
        if index.type == "text" and index.tokenizer.strip() == "":
            _push(
                issues,
                definition,
                "text_index_missing_tokenizer",
                f'Text index "{index.name}" on {definition.database}.{definition.name} '
                f"requires a tokenizer",
            )


def _validate_table(definition: TableDefinition, issues: list[ValidationIssue]) -> None:
    column_seen: set[str] = set()
    column_set: set[str] = set()
    for column in definition.columns:
        if column.name in column_seen:
            _push(
                issues,
                definition,
                "duplicate_column_name",
                f'Table {definition.database}.{definition.name} '
                f'has duplicate column name "{column.name}"',
            )
            continue
        column_seen.add(column.name)
        column_set.add(column.name)
        _validate_column_codec(definition, column, issues)

    _validate_indexes(definition, issues)

    projection_seen: set[str] = set()
    for projection in definition.projections or []:
        if projection.name in projection_seen:
            _push(
                issues,
                definition,
                "duplicate_projection_name",
                f'Table {definition.database}.{definition.name} '
                f'has duplicate projection name "{projection.name}"',
            )
            continue
        projection_seen.add(projection.name)

        # A projection carrying both keys satisfies the model, so Pydantic
        # admits it. Renders as index-only and drops the SELECT body on the
        # floor.
        if projection.index is not None and projection.query is not None:
            _push(
                issues,
                definition,
                "projection_ambiguous_kind",
                f'Table {definition.database}.{definition.name} projection '
                f'"{projection.name}" sets both "query" and "index"; use '
                f'"query" for a SELECT projection or "index"/"type" for an '
                f"index-only projection",
            )
            continue

        if (
            is_index_projection(projection)
            and normalize_projection_index(projection.index or "") == ""
        ):
            _push(
                issues,
                definition,
                "projection_empty_index",
                f'Table {definition.database}.{definition.name} projection '
                f'"{projection.name}" has an empty index expression',
            )

    for col in normalize_key_columns(definition.primary_key):
        if is_plain_column_reference(col) and col not in column_set:
            _push(
                issues,
                definition,
                "primary_key_missing_column",
                f"Table {definition.database}.{definition.name} primaryKey "
                f'references missing column "{col}"',
            )

    for col in normalize_key_columns(definition.order_by):
        if is_plain_column_reference(col) and col not in column_set:
            _push(
                issues,
                definition,
                "order_by_missing_column",
                f"Table {definition.database}.{definition.name} orderBy "
                f'references missing column "{col}"',
            )


_INTERVAL_PATTERN: Final[re.Pattern[str]] = re.compile(
    r"^\s*\d+\s+(SECOND|MINUTE|HOUR|DAY|WEEK|MONTH|YEAR)"
    r"(\s+\d+\s+(SECOND|MINUTE|HOUR|DAY|WEEK|MONTH|YEAR))*\s*$",
    re.IGNORECASE,
)

_REPLICATED_ENGINE_PATTERN: Final[re.Pattern[str]] = re.compile(r"^(Shared|Replicated)")


def _validate_interval(
    definition: MaterializedViewDefinition,
    issues: list[ValidationIssue],
    field: str,
    value: str | None,
) -> None:
    if value is None:
        return
    if _INTERVAL_PATTERN.match(value) is None:
        _push(
            issues,
            definition,
            "refresh_interval_format",
            f"Materialized view {definition.database}.{definition.name} "
            f'refresh.{field} "{value}" is not a valid interval '
            f'(expected e.g. "1 HOUR", "30 SECOND")',
        )


def _validate_materialized_view(
    definition: MaterializedViewDefinition,
    issues: list[ValidationIssue],
    definitions: list[SchemaDefinition],
) -> None:
    refresh = definition.refresh
    if refresh is None:
        return

    has_every = refresh.every is not None and len(refresh.every) > 0
    has_after = refresh.after is not None and len(refresh.after) > 0
    if not has_every and not has_after:
        _push(
            issues,
            definition,
            "refresh_requires_every_or_after",
            f"Materialized view {definition.database}.{definition.name} refresh "
            f'requires exactly one of "every" or "after"',
        )
    elif has_every and has_after:
        _push(
            issues,
            definition,
            "refresh_every_after_mutually_exclusive",
            f"Materialized view {definition.database}.{definition.name} refresh "
            f'specifies both "every" and "after"; choose one',
        )

    _validate_interval(definition, issues, "every", refresh.every)
    _validate_interval(definition, issues, "after", refresh.after)
    _validate_interval(definition, issues, "offset", refresh.offset)
    _validate_interval(definition, issues, "randomize", refresh.randomize)

    if (
        refresh.depends_on is not None
        and len(refresh.depends_on) > 0
        and has_after
        and not has_every
    ):
        _push(
            issues,
            definition,
            "refresh_depends_on_requires_every",
            f"Materialized view {definition.database}.{definition.name} uses "
            f"DEPENDS ON with REFRESH AFTER; ClickHouse only allows DEPENDS ON "
            f"with REFRESH EVERY.",
        )

    if not refresh.append:
        target: TableDefinition | None = None
        for other in definitions:
            if (
                isinstance(other, TableDefinition)
                and other.database == definition.to.database
                and other.name == definition.to.name
            ):
                target = other
                break
        if target is not None and _REPLICATED_ENGINE_PATTERN.match(target.engine) is not None:
            _push(
                issues,
                definition,
                "refresh_append_required_for_replicated_target",
                f"Materialized view {definition.database}.{definition.name} refreshes "
                f"a replicated target {target.database}.{target.name} ({target.engine}) "
                f"without APPEND. ClickHouse rejects this combination. Set "
                f"refresh.append = true, or target a non-replicated table.",
            )


def _validate_dictionary(  # noqa: PLR0912
    definition: DictionaryDefinition, issues: list[ValidationIssue]
) -> None:
    attribute_seen: set[str] = set()
    attribute_set: set[str] = set()
    for attribute in definition.attributes:
        if attribute.name in attribute_seen:
            _push(
                issues,
                definition,
                "duplicate_column_name",
                f"Dictionary {definition.database}.{definition.name} "
                f'has duplicate attribute name "{attribute.name}"',
            )
            continue
        attribute_seen.add(attribute.name)
        attribute_set.add(attribute.name)

        if attribute.default is not None and attribute.expression is not None:
            _push(
                issues,
                definition,
                "dictionary_attribute_default_expression_exclusive",
                f"Dictionary {definition.database}.{definition.name} attribute "
                f'"{attribute.name}" sets both "default" and "expression"; choose one',
            )

        if attribute.bidirectional and not attribute.hierarchical:
            _push(
                issues,
                definition,
                "dictionary_bidirectional_requires_hierarchical",
                f"Dictionary {definition.database}.{definition.name} attribute "
                f'"{attribute.name}" sets "bidirectional" without "hierarchical"; '
                f"bidirectional only applies to hierarchical attributes",
            )

    if len(definition.primary_key) == 0:
        _push(
            issues,
            definition,
            "dictionary_missing_primary_key",
            f"Dictionary {definition.database}.{definition.name} "
            f"requires a non-empty primaryKey",
        )
    else:
        for column in definition.primary_key:
            if column not in attribute_set:
                _push(
                    issues,
                    definition,
                    "dictionary_primary_key_missing_attribute",
                    f"Dictionary {definition.database}.{definition.name} primaryKey "
                    f'references missing attribute "{column}"',
                )

    if len(definition.source.strip()) == 0:
        _push(
            issues,
            definition,
            "dictionary_missing_source",
            f"Dictionary {definition.database}.{definition.name} "
            f'requires a non-empty "source"',
        )

    if len(definition.layout.strip()) == 0:
        _push(
            issues,
            definition,
            "dictionary_missing_layout",
            f"Dictionary {definition.database}.{definition.name} "
            f'requires a non-empty "layout"',
        )

    if len(definition.lifetime.strip()) == 0:
        _push(
            issues,
            definition,
            "dictionary_missing_lifetime",
            f"Dictionary {definition.database}.{definition.name} "
            f'requires a non-empty "lifetime"',
        )

    if definition.range is not None:
        for column in (definition.range.min, definition.range.max):
            if column not in attribute_set:
                _push(
                    issues,
                    definition,
                    "dictionary_range_missing_attribute",
                    f"Dictionary {definition.database}.{definition.name} range "
                    f'references missing attribute "{column}"',
                )


def validate_definitions(definitions: Iterable[SchemaDefinition]) -> list[ValidationIssue]:
    issues: list[ValidationIssue] = []
    object_keys: set[str] = set()
    materialized: list[SchemaDefinition] = list(definitions)
    for definition in materialized:
        key = definition_key(definition)
        if key in object_keys:
            _push(
                issues,
                definition,
                "duplicate_object_name",
                f'Duplicate schema object definition '
                f'"{definition.kind}:{definition.database}.{definition.name}"',
            )
            continue
        object_keys.add(key)

        if isinstance(definition, TableDefinition):
            _validate_table(definition, issues)
        elif isinstance(definition, MaterializedViewDefinition):
            _validate_materialized_view(definition, issues, materialized)
        elif isinstance(definition, DictionaryDefinition):
            _validate_dictionary(definition, issues)

    return issues


def assert_valid_definitions(definitions: Iterable[SchemaDefinition]) -> None:
    issues = validate_definitions(definitions)
    if len(issues) > 0:
        raise ChxValidationError(issues)

File: packages/core/src/validate.ts
import { definitionKey } from './canonical.js'
import { canonicalizeCodec, isGeneralCodec, isRawCodec } from './codec.js'
import { isPlainColumnReference, normalizeKeyColumns } from './key-clause.js'
import { isIndexProjection, normalizeProjectionIndex } from './projection.js'
import type {
  ColumnDefinition,
  DictionaryDefinition,
  MaterializedViewDefinition,
  MaterializedViewRefresh,
  SchemaDefinition,
  TableDefinition,
  ValidationIssue,
  ValidationIssueCode,
} from './model.js'
import { ChxValidationError as ValidationError } from './model.js'

function pushValidationIssue(
  issues: ValidationIssue[],
  def: SchemaDefinition,
  code: ValidationIssueCode,
  message: string
): void {
  issues.push({
    code,
    kind: def.kind,
    database: def.database,
    name: def.name,
    message,
  })
}

function validateColumnCodec(
  def: TableDefinition,
  column: ColumnDefinition,
  issues: ValidationIssue[]
): void {
  if (!column.codec) return
  const steps = canonicalizeCodec(column.codec)
  if (steps.length === 0) {
    pushValidationIssue(
      issues,
      def,
      'codec_chain_empty',
      `Table ${def.database}.${def.name} column "${column.name}" codec chain is empty; provide at least one codec or omit the field`
    )
    return
  }
  let generalCount = 0
  let generalIndex = -1
  for (const [i, step] of steps.entries()) {
    if (isRawCodec(step)) continue
    if (isGeneralCodec(step)) {
      generalCount += 1
      generalIndex = i
    }
  }

  if (generalCount > 1) {
    pushValidationIssue(
      issues,
      def,
      'codec_chain_multiple_general',
      `Table ${def.database}.${def.name} column "${column.name}" codec chain has more than one general codec; only one general codec is allowed at the end of a chain`
    )
    return
  }

  if (steps.length > 1 && generalCount === 1 && generalIndex !== steps.length - 1) {
    pushValidationIssue(
      issues,
      def,
      'codec_chain_must_end_with_general',
      `Table ${def.database}.${def.name} column "${column.name}" codec chain must end with a general codec (NONE, LZ4, LZ4HC, ZSTD, T64, GCD, ALP)`
    )
  }
}

function validateTableDefinition(def: TableDefinition, issues: ValidationIssue[]): void {
  const columnSeen = new Set<string>()
  const columnSet = new Set<string>()
  for (const column of def.columns) {
    if (columnSeen.has(column.name)) {
      pushValidationIssue(
        issues,
        def,
        'duplicate_column_name',
        `Table ${def.database}.${def.name} has duplicate column name "${column.name}"`
      )
      continue
    }
    columnSeen.add(column.name)
    columnSet.add(column.name)
    validateColumnCodec(def, column, issues)
  }

  const indexSeen = new Set<string>()
  for (const index of def.indexes ?? []) {
    if (indexSeen.has(index.name)) {
      pushValidationIssue(
        issues,
        def,
        'duplicate_index_name',
        `Table ${def.database}.${def.name} has duplicate index name "${index.name}"`
      )
      continue
    }
    indexSeen.add(index.name)
    if (index.type === 'text' && index.tokenizer.trim() === '') {
      pushValidationIssue(
        issues,
        def,
        'text_index_missing_tokenizer',
        `Text index "${index.name}" on ${def.database}.${def.name} requires a tokenizer`
      )
    }
  }

  const projectionSeen = new Set<string>()
  for (const projection of def.projections ?? []) {
    if (projectionSeen.has(projection.name)) {
      pushValidationIssue(
        issues,
        def,
        'duplicate_projection_name',
        `Table ${def.database}.${def.name} has duplicate projection name "${projection.name}"`
      )
      continue
    }
    projectionSeen.add(projection.name)

    // A projection carrying both keys satisfies the union, so TypeScript admits
    // it. Renders as index-only and drops the SELECT body on the floor.
    if ('index' in projection && 'query' in projection) {
      pushValidationIssue(
        issues,
        def,
        'projection_ambiguous_kind',
        `Table ${def.database}.${def.name} projection "${projection.name}" sets both "query" and "index"; use "query" for a SELECT projection or "index"/"type" for an index-only projection`
      )
      continue
    }

    if (isIndexProjection(projection) && normalizeProjectionIndex(projection.index) === '') {
      pushValidationIssue(
        issues,
        def,
        'projection_empty_index',
        `Table ${def.database}.${def.name} projection "${projection.name}" has an empty index expression`
      )
    }
  }

  for (const column of normalizeKeyColumns(def.primaryKey)) {
    if (isPlainColumnReference(column) && !columnSet.has(column)) {
      pushValidationIssue(
        issues,
        def,
        'primary_key_missing_column',
        `Table ${def.database}.${def.name} primaryKey references missing column "${column}"`
      )
    }
  }

  for (const column of normalizeKeyColumns(def.orderBy)) {
    if (isPlainColumnReference(column) && !columnSet.has(column)) {
      pushValidationIssue(
        issues,
        def,
        'order_by_missing_column',
        `Table ${def.database}.${def.name} orderBy references missing column "${column}"`
      )
    }
  }
}

const INTERVAL_PATTERN =
  /^\s*\d+\s+(SECOND|MINUTE|HOUR|DAY|WEEK|MONTH|YEAR)(\s+\d+\s+(SECOND|MINUTE|HOUR|DAY|WEEK|MONTH|YEAR))*\s*$/i

const REPLICATED_ENGINE_PATTERN = /^(Shared|Replicated)/

function validateInterval(
  def: MaterializedViewDefinition,
  issues: ValidationIssue[],
  field: keyof MaterializedViewRefresh,
  value: string | undefined
): void {
  if (value === undefined) return
  if (!INTERVAL_PATTERN.test(value)) {
    pushValidationIssue(
      issues,
      def,
      'refresh_interval_format',
      `Materialized view ${def.database}.${def.name} refresh.${String(field)} "${value}" is not a valid interval (expected e.g. "1 HOUR", "30 SECOND")`
    )
  }
}

function validateMaterializedViewDefinition(
  def: MaterializedViewDefinition,
  issues: ValidationIssue[],
  definitions: SchemaDefinition[]
): void {
  const { refresh } = def
  if (!refresh) return

  const hasEvery = refresh.every !== undefined && refresh.every.length > 0
  const hasAfter = refresh.after !== undefined && refresh.after.length > 0
  if (!hasEvery && !hasAfter) {
    pushValidationIssue(
      issues,
      def,
      'refresh_requires_every_or_after',
      `Materialized view ${def.database}.${def.name} refresh requires exactly one of "every" or "after"`
    )
  } else if (hasEvery && hasAfter) {
    pushValidationIssue(
      issues,
      def,
      'refresh_every_after_mutually_exclusive',
      `Materialized view ${def.database}.${def.name} refresh specifies both "every" and "after"; choose one`
    )
  }

  validateInterval(def, issues, 'every', refresh.every)
  validateInterval(def, issues, 'after', refresh.after)
  validateInterval(def, issues, 'offset', refresh.offset)
  validateInterval(def, issues, 'randomize', refresh.randomize)

  if (refresh.dependsOn && refresh.dependsOn.length > 0 && hasAfter && !hasEvery) {
    pushValidationIssue(
      issues,
      def,
      'refresh_depends_on_requires_every',
      `Materialized view ${def.database}.${def.name} uses DEPENDS ON with REFRESH AFTER; ClickHouse only allows DEPENDS ON with REFRESH EVERY.`
    )
  }

  if (!refresh.append) {
    const target = definitions.find(
      (other): other is TableDefinition =>
        other.kind === 'table' &&
        other.database === def.to.database &&
        other.name === def.to.name
    )
    if (target && REPLICATED_ENGINE_PATTERN.test(target.engine)) {
      pushValidationIssue(
        issues,
        def,
        'refresh_append_required_for_replicated_target',
        `Materialized view ${def.database}.${def.name} refreshes a replicated target ${target.database}.${target.name} (${target.engine}) without APPEND. ClickHouse rejects this combination. Set refresh.append = true, or target a non-replicated table.`
      )
    }
  }
}

function validateDictionaryDefinition(def: DictionaryDefinition, issues: ValidationIssue[]): void {
  const attributeSeen = new Set<string>()
  const attributeSet = new Set<string>()
  for (const attribute of def.attributes) {
    if (attributeSeen.has(attribute.name)) {
      pushValidationIssue(
        issues,
        def,
        'duplicate_column_name',
        `Dictionary ${def.database}.${def.name} has duplicate attribute name "${attribute.name}"`
      )
      continue
    }
    attributeSeen.add(attribute.name)
    attributeSet.add(attribute.name)

    if (attribute.default !== undefined && attribute.expression !== undefined) {
      pushValidationIssue(
        issues,
        def,
        'dictionary_attribute_default_expression_exclusive',
        `Dictionary ${def.database}.${def.name} attribute "${attribute.name}" sets both "default" and "expression"; choose one`
      )
    }

    if (attribute.bidirectional && !attribute.hierarchical) {
      pushValidationIssue(
        issues,
        def,
        'dictionary_bidirectional_requires_hierarchical',
        `Dictionary ${def.database}.${def.name} attribute "${attribute.name}" sets "bidirectional" without "hierarchical"; bidirectional only applies to hierarchical attributes`
      )
    }
  }

  if (def.primaryKey.length === 0) {
    pushValidationIssue(
      issues,
      def,
      'dictionary_missing_primary_key',
      `Dictionary ${def.database}.${def.name} requires a non-empty primaryKey`
    )
  } else {
    for (const column of def.primaryKey) {
      if (!attributeSet.has(column)) {
        pushValidationIssue(
          issues,
          def,
          'dictionary_primary_key_missing_attribute',
          `Dictionary ${def.database}.${def.name} primaryKey references missing attribute "${column}"`
        )
      }
    }
  }

  if (def.source.trim().length === 0) {
    pushValidationIssue(
      issues,
      def,
      'dictionary_missing_source',
      `Dictionary ${def.database}.${def.name} requires a non-empty "source"`
    )
  }

  if (def.layout.trim().length === 0) {
    pushValidationIssue(
      issues,
      def,
      'dictionary_missing_layout',
      `Dictionary ${def.database}.${def.name} requires a non-empty "layout"`
    )
  }

  if (def.lifetime.trim().length === 0) {
    pushValidationIssue(
      issues,
      def,
      'dictionary_missing_lifetime',
      `Dictionary ${def.database}.${def.name} requires a non-empty "lifetime"`
    )
  }

  if (def.range) {
    for (const column of [def.range.min, def.range.max]) {
      if (!attributeSet.has(column)) {
        pushValidationIssue(
          issues,
          def,
          'dictionary_range_missing_attribute',
          `Dictionary ${def.database}.${def.name} range references missing attribute "${column}"`
        )
      }
    }
  }
}

export function validateDefinitions(definitions: SchemaDefinition[]): ValidationIssue[] {
  const issues: ValidationIssue[] = []
  const objectKeys = new Set<string>()
  for (const def of definitions) {
    const key = definitionKey(def)
    if (objectKeys.has(key)) {
      pushValidationIssue(
        issues,
        def,
        'duplicate_object_name',
        `Duplicate schema object definition "${def.kind}:${def.database}.${def.name}"`
      )
      continue
    }
    objectKeys.add(key)

    if (def.kind === 'table') {
      validateTableDefinition(def, issues)
    } else if (def.kind === 'materialized_view') {
      validateMaterializedViewDefinition(def, issues, definitions)
    } else if (def.kind === 'dictionary') {
      validateDictionaryDefinition(def, issues)
    }
  }

  return issues
}

export function assertValidDefinitions(definitions: SchemaDefinition[]): void {
  const issues = validateDefinitions(definitions)
  if (issues.length > 0) throw new ValidationError(issues)
}
